File: src/pneuma_core/models/goals.py
"""Goal hierarchy models: Vision, Objective, Task, GoalTree."""

from dataclasses import dataclass, field
from typing import Literal

ObjectiveStatus = Literal["active", "achieved", "abandoned"]
TaskStatus = Literal["pending", "in_progress", "completed", "abandoned"]

_OBJECTIVE_STATUSES = ("active", "achieved", "abandoned")
_TASK_STATUSES = ("pending", "in_progress", "completed", "abandoned")


@dataclass(frozen=True)
class Vision:
    """長期ビジョン（5-10年）."""

    id: str
    character_id: str
    content: str


@dataclass(frozen=True)
class Objective:
    """中期目標（四半期〜年単位）.

    status: active / achieved / abandoned
    progress: 0.0〜1.0
    """

    id: str
    character_id: str
    vision_id: str
    content: str
    status: ObjectiveStatus
    progress: float

    def __post_init__(self) -> None:
        if self.status not in _OBJECTIVE_STATUSES:
            raise ValueError(
                f"status must be one of {_OBJECTIVE_STATUSES}, got '{self.status}'"
            )
        if not (0.0 <= self.progress <= 1.0):
            raise ValueError(f"progress must be between 0.0 and 1.0, got {self.progress}")


@dataclass(frozen=True)
class Task:
    """具体的タスク（日〜週単位）.

    status: pending / in_progress / completed / abandoned
    """

    id: str
    character_id: str
    objective_id: str
    content: str
    status: TaskStatus

    def __post_init__(self) -> None:
        if self.status not in _TASK_STATUSES:
            raise ValueError(
                f"status must be one of {_TASK_STATUSES}, got '{self.status}'"
            )


@dataclass
class GoalTree:
    """3階層の目標構造.

    Vision → Objective → Task の階層参照を提供する。
    """

    visions: list[Vision] = field(default_factory=list)
    objectives: list[Objective] = field(default_factory=list)
    tasks: list[Task] = field(default_factory=list)

    def get_objectives_for_vision(self, vision_id: str) -> list[Objective]:
        """Vision に紐づく Objectives を取得."""
        return [o for o in self.objectives if o.vision_id == vision_id]

    def get_tasks_for_objective(self, objective_id: str) -> list[Task]:
        """Objective に紐づく Tasks を取得."""
        return [t for t in self.tasks if t.objective_id == objective_id]

File: akasha/contracts/Goals.ts
// Akasha Contract — Goals (Vision / Objective / Task / GoalTree)
//
// Python 一次ソース: src/pneuma_core/models/goals.py
//
// 3 階層構造を「フラットなリスト + ID 参照」で持つ。
// 階層的に nest はしない (`get_objectives_for_vision()` 等で関連付ける)。

export type ObjectiveStatus = "active" | "achieved" | "abandoned";
export type TaskStatus =
  | "pending"
  | "in_progress"
  | "completed"
  | "abandoned";

export interface Vision {
  id: string;
  character_id: string;
  /** 5–10 年スパンの長期ビジョン。 */
  content: string;
}

export interface Objective {
  id: string;
  character_id: string;
  vision_id: string;
  content: string;
  status: ObjectiveStatus;
  /** [0, 1]。 */
  progress: number;
}

export interface Task {
  id: string;
  character_id: string;
  objective_id: string;
  content: string;
  status: TaskStatus;
}

/** 3 階層の goal を保持する mutable コンテナ。 */
export interface GoalTree {
  visions: Vision[];
  objectives: Objective[];
  tasks: Task[];
}
